bindings/rust/bench/src/harness.rs

```diff
@@ -1,163 +1,356 @@
+        Self::new(
+            mode,
+            CryptoConfig::default(),
+            HandshakeType::default(),
+            ConnectedBuffer::default(),
+        )
     }
 
+    /// Make a config with given parameters
+    fn make_config(
+        mode: Mode,
+        crypto_config: CryptoConfig,
+        handshake_type: HandshakeType,
+    ) -> Result<Self::Config, Box<dyn Error>>;
+
+    /// Make connection from existing config and buffer
+    fn new_from_config(
+        config: &Self::Config,
+        connected_buffer: ConnectedBuffer,
+    ) -> Result<Self, Box<dyn Error>>;
+
     /// Initialize buffers, configs, and connections (pre-handshake)
     fn new(
+        mode: Mode,
         crypto_config: CryptoConfig,
         handshake_type: HandshakeType,
         buffer: ConnectedBuffer,
-    ) -> Result<Self, Box<dyn Error>>;
+    ) -> Result<Self, Box<dyn Error>> {
+        Self::new_from_config(
+            &Self::make_config(mode, crypto_config, handshake_type)?,
+            buffer,
+        )
+    }
 
-    /// Run handshake on initialized connections
-    /// Returns error if handshake has already completed
+    /// Run one handshake step: receive msgs from other connection, process, and send new msgs
     fn handshake(&mut self) -> Result<(), Box<dyn Error>>;
 
-    /// Checks if handshake is finished for both client and server
     fn handshake_completed(&self) -> bool;
 
-    /// Get negotiated cipher suite
     fn get_negotiated_cipher_suite(&self) -> CipherSuite;
 
-    /// Get whether or negotiated version is TLS1.3
     fn negotiated_tls13(&self) -> bool;
 
-    /// Send application data from connection in harness pair
-    fn send(&mut self, sender: Mode, data: &[u8]) -> Result<(), Box<dyn Error>>;
+    /// Send application data to ConnectedBuffer
+    fn send(&mut self, data: &[u8]) -> Result<(), Box<dyn Error>>;
+
+    /// Read application data from ConnectedBuffer
+    fn recv(&mut self, data: &mut [u8]) -> Result<(), Box<dyn Error>>;
+
+    /// Shrink buffers owned by the connection
+    fn shrink_connection_buffers(&mut self);
 
-    /// Receive application data sent to connection in harness pair
-    fn recv(&mut self, receiver: Mode, data: &mut [u8]) -> Result<(), Box<dyn Error>>;
+    /// Clear and shrink buffers used for IO with another connection
+    fn shrink_connected_buffer(&mut self);
 
-    /// Send data from client to server and then from server to client
-    fn round_trip_transfer(&mut self, data: &mut [u8]) -> Result<(), Box<dyn Error>> {
+    /// Get reference to internal connected buffer
+    fn connected_buffer(&self) -> &ConnectedBuffer;
+}
+
+pub struct TlsConnPair<C: TlsConnection, S: TlsConnection> {
+    client: C,
+    server: S,
+}
+
+impl<C: TlsConnection, S: TlsConnection> Default for TlsConnPair<C, S> {
+    fn default() -> Self {
+        Self::new(Default::default(), Default::default(), Default::default()).unwrap()
+    }
+}
+
+impl<C: TlsConnection, S: TlsConnection> TlsConnPair<C, S> {
```

Comment:
Can I ask, in general, what purpose does it serve to separate out the client and server connections?

>  a Rustls client could be tested with a s2n-tls server.

Why would this test be useful for benchmarking?

Reply:
It does let us do interop testing, which is pretty cool.

Reply:
Sure, but this is benchmarking code, we're not using this for interop testing. I don't know the official project plan, but doing this kind of out of scope for this project?

Reply:
I don't want to say it's out of scope, since the purpose for it for benching is to provide a way to make sure that connection pairs that are mixed library types have the same performance for each individual connection when compared to their respective connection pairs with the same library type.

Reply:
I guess, I think this is a good change to have this code be more flexible, but I don't actually think a Rustls to s2n-tls handshake would produce more/better benchmarking information than a Rustls to Rustls handshake or an s2n to s2n handshake.